### PR TITLE
Enable external tools for the vertical bar, move implementation out of the code-behind

### LINKED ISSUE #3043: The "Pin" button for an external tool in the context menu has the wrong icon (shows UnPin).
### Dev Home version

0.503

### Windows build number

10.0.22631.3672

### Other software

_No response_

### Steps to reproduce the bug

Right-click to get external tools menu

### Expected result

Pin shows pin glyph, Unpin shows unpin glyph

### Actual result

Pin shows unpin glyph, Unpin shows pin glyph

### Included System Information

_No response_

### Included Extensions Information

_No response_

### PR DESCRIPTION
**Pull request by timkur**
## Summary of the pull request

This pulls all the external tool support into the BarViewModel, and lights up the functionality to work for both the horizontal and vertical bar.

## References and relevant issues

A few updates to the initial PR

WinUI3 will display an empty menu flyout when it doesn't have any items... ideally we don't want an empty flyout to show at all with 0 items, so the viewmodel will now create/destroy a menuflyout based on if we have items or not.

This also fixes an issue with the navigation to settings feature when we weren't attached to a process. The code that did a "set this to either 0 or Items-1" was always set to 0, causing navigation issues.

Last but not least, I was getting some random "invalid args" when trying to read images. Updated the code in ExternalTools to no longer rely on temp files.

## PR checklist
- [ ] Closes #3043, #3041
- [ ] Tests added/passed
- [ ] Documentation updated

### CODE REVIEW

File: tools/PI/DevHome.PI/BarWindowVertical.xaml [hunk 1]
```diff
@@ -81,9 +81,10 @@
             </Button>
 
             <Button
-                x:Name="ManageExternalToolsButton"
-                x:Uid="ManageExternalToolsButton"
-                HorizontalAlignment="Center">
+                 x:Name="ManageExternalToolsButton"
+                 x:Uid="ManageExternalToolsButton"
+                 HorizontalAlignment="Center"
+                 Click="{x:Bind _viewModel.ManageExternalToolsButton_Click}">
```
Comment (krschau) [inline tools/PI/DevHome.PI/BarWindowVertical.xaml:87]: Nit: indentation is off

Comment (timkur) [reply tools/PI/DevHome.PI/BarWindowVertical.xaml]: Thanks.

File: tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs [hunk 2]
```diff
@@ -236,4 +272,255 @@ private void PerfCounterHelper_PropertyChanged(object? sender, PropertyChangedEv
...
+    }
+
+    // This creates a Menu Flyout item for an external tool. It also creates a sub-menu item for pinning/unpinning
+    // and unregistering the tool.
+    private MenuFlyoutItem CreateContextMenuItemForTool(ExternalTool tool)
+    {
+        var imageIcon = new ImageIcon
+        {
+            Source = tool.ToolIcon,
+        };
```
Comment (zadesai) [inline tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs:348]: Is it possible to create a MenuFlyoutItemControl and use it here?

Comment (timkur) [reply tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs:348]: Can you elaborate?

Comment (zadesai) [reply tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs:348]: Sure, can work on it once this PR is in.

**Review by zadesai — APPROVED** [review]
:shipit:

File: tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs (hunk 2 above)
Comment (timkur) [reply tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs]: Makes sense, and it did simplify some things as well.

**Review by zadesai — APPROVED** [review]
:shipit:

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    }
+
+    private void AddExternalToolToContextMenu(MenuFlyout menuFlyout, ExternalTool tool)
+    {
+        menuFlyout.Items.Add(CreateContextMenuItemForTool(tool));
+    }
+
+    private void RemoveExternalToolFromContextMenu(MenuFlyout menuFlyout, ExternalTool tool)
+    {
+        foreach (var menuItem in menuFlyout.Items)
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:108]: NIT: `menuFlyout.Items.Where(x => x.Tag == tool)`

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    private void ExternalTools_CollectionChanged(object? sender, NotifyCollectionChangedEventArgs e)
+    {
+        _dispatcher.TryEnqueue(() =>
+        {
+            if (e.Action == NotifyCollectionChangedAction.Add && e.NewItems is not null)
+            {
+                foreach (ExternalTool tool in e.NewItems)
+                {
+                    // If this button doens't have a flyout, create one
+                    if (this.ContextFlyout is not MenuFlyout flyout)
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:67]: I don't understand why ContextFlyout isn't a MenuFlyout is a signal that the button does not have a flyout.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        menuItem.ContextFlyout = menuSubItemFlyout;
+
+        return menuItem;
+    }
+
+    private void ExternalToolItem_PropertyChanged(object? sender, PropertyChangedEventArgs e)
+    {
+        if (sender is not ExternalTool tool)
+        {
+            Debug.Assert(false, "What is this?");
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:184]: Please use a more descriptive debug message.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+            return;
+        }
+
+        _dispatcher.TryEnqueue(() =>
+        {
+            MenuFlyout? flyout = this.ContextFlyout as MenuFlyout;
+            Debug.Assert(flyout is not null, "Why does this button not have a menu flyout?");
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:191]: If the ContextMenu is needs to be a MenuFlyout, please move this check to the top so this check can be performed earlier and future readers can notice this faster.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+                    var imageIcon = new ImageIcon
+                    {
+                        Source = tool.ToolIcon,
+                    };
+
+                    menuItem.Icon = imageIcon;
+
+                    var menuSubItemFlyout = menuItem.ContextFlyout as MenuFlyout;
+                    Debug.Assert(menuSubItemFlyout != null, "Why is this null?");
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:209]: Please user more descriptive messages.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+
+                    var pinSubItemItem = menuSubItemFlyout.Items[0] as MenuFlyoutItem;
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:211]: .Items[0] can crash if empty.  Consider using ?[0]

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+                    Debug.Assert(pinSubItemItem != null, "Why is this null?");
+
+                    var unPinSubItemItem = menuSubItemFlyout.Items[1] as MenuFlyoutItem;
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:214]: Can crash if Items is null, or length == 1.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+
+    public void ExternalToolMenuItem_Click(object sender, RoutedEventArgs e)
+    {
+        if (sender is MenuFlyoutItem clickedMenuItem)
+        {
+            if (clickedMenuItem.Tag is ExternalTool tool)
+            {
+                ExternalToolLaunchRequest?.Invoke(this, tool);
+
+                // InvokeTool(tool, TargetAppData.Instance.TargetProcess?.Id, TargetAppData.Instance.HWnd);
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:233]: Remove commented out code.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+                    unPinSubItemItem.Visibility = tool.IsPinned ? Visibility.Visible : Visibility.Collapsed;
+                    break;
+                }
+            }
+        });
+    }
+
+    public void ExternalToolMenuItem_Click(object sender, RoutedEventArgs e)
+    {
+        if (sender is MenuFlyoutItem clickedMenuItem)
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:227]: NIT: `sender is MenuFlyoutItem clickedMenuItem && clickedMenuItemTag is ExternalTool tool`

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+        ExternalTool tool = GetToolFromSender(sender);
+        tool.UnregisterTool();
+
+        HideFlyout(sender);
+    }
+
+    private ExternalTool GetToolFromSender(object sender)
+    {
+        MenuFlyoutItem? clickedMenuItem = sender as MenuFlyoutItem;
+        Debug.Assert(clickedMenuItem != null, "Why is this null?");
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:256]: Please use a descriptive error message.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+
+        ExternalTool? tool = clickedMenuItem.Tag as ExternalTool;
+        Debug.Assert(tool != null, "Why is this null?");
+
+        return tool;
+    }
+
+    private void HideFlyout(object sender)
+    {
+        if (this.ContextFlyout is MenuFlyout flyout)
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:266]: So much of this code casts `ContextFlyout` to `MenuItem` why?  Does the ContextFlyout change what type it is through out the application's lifetime?

File: tools/PI/DevHome.PI/Helpers/ExternalTool.cs [hunk 14]
```diff
@@ -75,50 +78,39 @@ public string OtherArgs
...
         if (!string.IsNullOrEmpty(executable))
         {
-            GetToolImage();
-            GetMenuIcon();
+            GetIcons();
         }
     }
 
-    private async void GetToolImage()
+    partial void OnIsPinnedChanged(bool oldValue, bool newValue)
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Helpers/ExternalTool.cs:91]: What does `oldValue` and `newValue`` represent?

File: tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs [hunk 15]
```diff
@@ -168,6 +173,18 @@ public void ProcessChooserCommand()
         barWindow?.NavigateTo(typeof(ProcessListPageViewModel));
     }
 
+    public void ManageExternalToolsButton_Click(object sender, RoutedEventArgs e)
+    {
+        // Need to be in a horizontal layout
+        BarOrientation = Orientation.Horizontal;
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs:179]: Why?

File: tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs [hunk 16]
```diff
@@ -236,4 +253,35 @@ private void PerfCounterHelper_PropertyChanged(object? sender, PropertyChangedEv
...
+    }
+
+    public void ManageExternalToolsButton_ExternalToolLaunchRequest(object sender, ExternalTool tool)
+    {
+        InvokeTool(tool, TargetAppData.Instance.TargetProcess?.Id, TargetAppData.Instance.HWnd);
+    }
+
+    private void InvokeTool(ExternalTool tool, int? id, HWND hWnd)
+    {
+        var process = tool.Invoke(id, hWnd);
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs:275]: will tool.Invoke throw if id is null?

File: tools/PI/DevHome.PI/Helpers/ExternalTool.cs [hunk 17]
```diff
@@ -75,50 +78,39 @@ public string OtherArgs
...
             _softwareBitmap ??= GetSoftwareBitmapFromExecutable(Executable);
             if (_softwareBitmap is not null)
             {
-                var bitmapUri = await SaveSoftwareBitmapToTempFile(_softwareBitmap);
-                MenuIcon = new BitmapIcon
-                {
-                    UriSource = bitmapUri,
-                    ShowAsMonochrome = false,
-                };
+                ToolIcon = await GetSoftwareBitmapSourceFromSoftwareBitmap(_softwareBitmap);
```
Comment (dhoehna) [inline tools/PI/DevHome.PI/Helpers/ExternalTool.cs:103]: add `async` to the end of the method name.

File: tools/PI/DevHome.PI/BarWindowHorizontal.xaml [hunk 18]
```diff
@@ -77,38 +77,33 @@
...
                 x:Uid="ManageExternalToolsButton"
                 HorizontalAlignment="Center"
-                Click="ManageExternalToolsButton_Click">
-                <Button.ContextFlyout>
-                    <MenuFlyout x:Name="ExternalToolsMenu">
-                        <!-- We can't databind to MenuFlyout, so we'll add a MenuFlyoutItem
-                        for each external tool in codebehind here. -->
-                    </MenuFlyout>
-                </Button.ContextFlyout>
+                Click="{x:Bind _viewModel.ManageExternalToolsButton_Click}"
```
[the remark this answers is not in the document]
Comment (AmelBawa-msft) [reply tools/PI/DevHome.PI/BarWindowHorizontal.xaml:84]: For MVVM I suggest using commands instead of event handlers. There are several benefits for commands including decoupling the UI classes from the view-model. For example, you won't need to depend on `RoutedEventArgs` from the view-model and instead you can customize your own params. Commands are also reusable by multiple controls (again since we control the parameter and don't depend on a particular signature of a specific control). Now more specifically, the RelayCommand attribute from the MVVM toolkit it also allows controlling execution concurrency and has the option to bind the `CanExecute` to enable/disable a control.

TLDR; For UI manipulation (e.g. a button click that expands/collapses another control) I would use a regular event handler bound to a code-behind method. For everything else, I'd consider adding a `RelayCommand` in the view-model.

Related links:
- [RelayCommand attribute](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/generators/relaycommand)

File: tools/PI/DevHome.PI/BarWindow.cs [hunk 19]
```diff
@@ -76,6 +76,8 @@ internal void SetRequestedTheme(ElementTheme theme)
 
     public void NavigateTo(Type viewModelType) => _horizontalWindow.NavigateTo(viewModelType);
 
+    public void NavigateToSettings(string settingsPage) => _horizontalWindow.NavigateToSettings(settingsPage);
```
Comment (krschau) [inline tools/PI/DevHome.PI/BarWindow.cs:79]: Reading this I assumed it meant Dev Home settings, but on closer inspection it seemed to be PI-specific settings. Consider making that more clear in the method and/or parameter names.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 13 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:266]: The cast is not needed in this particular function... so I'll remove it.

But in general, the code is programmatically building a MenuFlyout and manipulating the MenuFlyoutItems based on the state of the external tools.

WinUI3 accepts this MenuFlyout as a FlyoutBase object (stored in ContextFlyout). I can't get MenuFlyoutItems directly from a FlyoutBase, which is why there is the cast to a MenuFlyout.

Does that answer your question?

File: tools/PI/DevHome.PI/BarWindow.cs (hunk 19 above)
Comment (timkur) [reply tools/PI/DevHome.PI/BarWindow.cs:79]: Changed to NavigateToPiSettings

File: tools/PI/DevHome.PI/Helpers/ExternalTool.cs (hunk 17 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Helpers/ExternalTool.cs:103]: Ok.

File: tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs (hunk 16 above)
Comment (timkur) [reply tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs:275]: Nope. This tells the external tool that we don't have a Process ID to pass to the tool. In this case, "0" and "null" have 2 different meanings.

Comment (timkur) [reply tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs:275]: I changed the param name to pid to make it more clear what sort of ID this is.

File: tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs (hunk 15 above)
Comment (timkur) [reply tools/PI/DevHome.PI/ViewModels/BarWindowViewModel.cs:179]: Our expanded content is only shown in the horizontal position. I'll update the comment.

File: tools/PI/DevHome.PI/Helpers/ExternalTool.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (timkur) [reply tools/PI/DevHome.PI/Helpers/ExternalTool.cs:91]: These are generated as part of the [ObservableProperty]. These allow you to take action on these properties being changed in your view model.  

[https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/generators/observableproperty#running-code-upon-changes](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/generators/observableproperty#running-code-upon-changes)

Comment (timkur) [reply tools/PI/DevHome.PI/Helpers/ExternalTool.cs:91]: For 

 [ObservableProperty]
 private bool _isPinned;

It's what the old value of _isPinned was, and what the new value is when it gets changed.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 10 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:233]: Thanks.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 11 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:227]: Thanks.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 8 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:211]: I'd prefer if it crashed. It means the menu flyout wasn't created properly.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 9 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:214]: I'd prefer if it crashed. It means the menu flyout wasn't created properly.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 7 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:209]: Ok. Gone are the days of ASSERT(p != NULL). :)

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 12 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:256]: Done.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+            return;
+        }
+
+        _dispatcher.TryEnqueue(() =>
+        {
+            MenuFlyout? flyout = this.ContextFlyout as MenuFlyout;
+            Debug.Assert(flyout is not null, "Why does this button not have a menu flyout?");
+
+            // Update the submenu item for this tool
+            foreach (MenuFlyoutItem menuItem in flyout.Items)
```
[the remark this answers is not in the document]
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:194]: Thanks!

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 3 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:108]: Thanks!

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 6 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:191]: I wanted to perform the check on the UI thread, not only to avoid race conditions, the just the general goodness of accessing UI elements on the UI thread.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 5 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:184]: Ok.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 4 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:67]: Could have done

if (this.ContextFlyout is null)
...

MenuFlyout flyout = this.ContextFlyout as MenuFlyout;
Debug.Assert(flyout is not null);

But this seemed more compact and accomplished what I was looking for.

File: tools/PI/DevHome.PI/BarWindowHorizontal.xaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (timkur) [reply tools/PI/DevHome.PI/BarWindowHorizontal.xaml:84]: I moved this particular instance to command binding. Ideally I would have also done it for the one below (_viewModel.ExternalToolButton_Click") but things get funky when you're in a datatemplate and it's not super straightward. I'll save moving that one for a rainy day.

File: tools/PI/DevHome.PI/ViewModels/ExpandedViewControlViewModel.cs [hunk 21]
```diff
@@ -236,23 +236,19 @@ public void NavigateTo(Type viewModelType)
...
         // we need to move the list selection so that when they subsequently select an item from
         // the NavLinks, we'll navigate to the correct page even if that was the previously-selected item.
-        if (SelectedNavLinkIndex == 0)
-        {
-            SelectedNavLinkIndex = Links.Count - 1;
-        }
-        else
-        {
-            SelectedNavLinkIndex = 0;
-        }
```
Comment (andreww-msft) [inline tools/PI/DevHome.PI/ViewModels/ExpandedViewControlViewModel.cs:255]: Hmm... I tried this before, but it throws in some scenarios.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs [hunk 22]
```diff
@@ -0,0 +1,267 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
```
Comment (andreww-msft) [inline tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:10]: What are we using this for?

File: tools/PI/DevHome.PI/ViewModels/ExpandedViewControlViewModel.cs (hunk 21 above)
Comment (timkur) [reply tools/PI/DevHome.PI/ViewModels/ExpandedViewControlViewModel.cs:255]: Do you remember what scenarios? After adding the -1 check in Navigate() above, I haven't seen any issues.

The code as-is is broken in the case where we're not attached to any process, because Links.Count=1... so we're always setting SelectedNavLinkIndex to 0.

Not only does it jam the UI (for the reasons described above) but setting this to 0 (even if SelectedNavLinkIndex is already 0) ends up triggering another Navigate call, and so after navigating to the settings page we immediately navigate to page 0, navigating away from the settings page.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 22 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs]: They snuck in. Removed these.

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 4 above)
Comment (dhoehna) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:61]: What if `ContextMenu` is not null and not a `MenuFlyout`?  Also, if I am confused about the meaning of the comment, other people will be as well.  Consider adding more description explaining why something that isn't a `MenuFlyout` equates to "The button does not have a flyout"

File: tools/PI/DevHome.PI/Helpers/ExternalTool.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (dhoehna) [reply tools/PI/DevHome.PI/Helpers/ExternalTool.cs:91]: Oooohhh.  I see yeah.  I understand.  I did not override any methods from the generated partial class.  The name went over my head.

**Review by dhoehna — APPROVED** [review]

File: tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs (hunk 4 above)
Comment (timkur) [reply tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs:61]: If that's the case, then bad things have happened, meaning that somebody else is mucking with this control's flyout. I could add an assert that nobody else is mucking with my context menu...
